File: biobakery_workflows/scripts/learn_error_rates.R
#!/usr/bin/env Rscript

# load packages
library(dada2); packageVersion("dada2")
library(ggplot2)

## Collect arguments
args <- commandArgs(TRUE)

## Parse arguments (we expect the form --arg=value)
parseArgs <- function(x) strsplit(sub("^--", "", x), "=")
args.df <- as.data.frame(do.call("rbind", parseArgs(args)))

args.list <- as.list(as.character(args.df$V2))
names(args.list) <- args.df$V1

## Arg1 default
if(is.null(args.list$output_dir)) {
  stop("At least one argument must be supplied (output folder).\n", call.=FALSE)
}

# Print args list to STDOUT
for( i in names(args.list) ) {
  cat( i, "\t", args.list[[i]], "\n")
}

# print contents of folder
cat( grep( "*\\.fastq.gz", list.files(args.list$output_dir), value=T ), sep = "\n" )

# these variables are passed to the workflow
output.path <- normalizePath( args.list$output_dir )
print(output.path)

filt.path = file.path(output.path,"filtered_input")

filtFs <- file.path(filt.path,sort(grep( "*_F_filt.fastq*", list.files(filt.path), value = T ) ))
filtRs <- file.path(filt.path,sort(grep( "*_R_filt.fastq*", list.files(filt.path), value = T ) ))

set.seed(100)
# Filtered forward reads
errF <- dada2::learnErrors(filtFs, nread=1e3, multithread=TRUE)
# Filtered reverse reads
errR <- dada2::learnErrors(filtRs, nread=1e3, multithread=TRUE)


# Visualize the estimated error rates
ggplot2::ggsave(paste0(output.path,"/Error_rates_per_sample_FWD.png"), dada2::plotErrors(errF, nominalQ=TRUE) , device = "png")
ggplot2::ggsave(paste0(output.path,"/Error_rates_per_sample_REV.png"), dada2::plotErrors(errR, nominalQ=TRUE) , device = "png")

#save as rds files
saveRDS(errF, paste0(output.path, "/error_rates_F.rds")) 
saveRDS(errR, paste0(output.path, "/error_rates_R.rds"))

File: biobakery_workflows/scripts/learn_error_rates.py
#!/usr/bin/env python

from rpy2 import robjects
from rpy2.robjects.packages import importr
import os.path
import sys
import re
import argparse

base = importr('base')
dada2 = importr('dada2')
ggplot2 = importr('ggplot2')

def parse_arguments(args):
    """ Parse the arguments"""

    parser = argparse.ArgumentParser(
        description="Learn error rates for each sample\n",
        formatter_class=argparse.RawTextHelpFormatter)

    parser.add_argument(
        "--filt_path",
        help="path to filtered files \n[REQUIRED]",
    )
    parser.add_argument(
        "--error_ratesF_png",
        help="error rates forward reads plot\n[REQUIRED]",
    )
    parser.add_argument(
        "--error_ratesR_png",
        help="error rates reverse reads plot\n[REQUIRED]",
    )
    parser.add_argument(
        "--error_ratesF_rds",
        help="error rates forward reads data\n[REQUIRED]",
    )
    parser.add_argument(
        "--error_ratesR_rds",
        help = "error rates reverse reads data\n[REQUIRED]",
    )
    parser.add_argument(
        "--threads",
        help="number of threads\n[REQUIRED]",
    )

    return parser.parse_args()


def main():

    # parse arguments from the user
    args = parse_arguments(sys.argv)

    filtFs = sorted(
        [os.path.join(args.filt_path, f) for f in os.listdir(args.filt_path) if re.match(r'(.*)_F_filt.fastq(.*)', f)])
    filtRs = sorted(
        [os.path.join(args.filt_path, f) for f in os.listdir(args.filt_path) if re.match(r'(.*)_R_filt.fastq(.*)', f)])

    robjects.globalenv["filtFs"] = filtFs
    robjects.globalenv["filtRs"] = filtRs
    robjects.globalenv["error_ratesF_png"] = args.error_ratesF_png
    robjects.globalenv["error_ratesR_png"] = args.error_ratesR_png
    robjects.globalenv["error_ratesF_rds"] = args.error_ratesF_rds
    robjects.globalenv["error_ratesR_rds"] = args.error_ratesR_rds
    robjects.globalenv["threads"] = args.threads


    robjects.r('''

        filtF <- unlist(filtFs, use.names=FALSE)
        filtR <- unlist(filtRs, use.names=FALSE)

        set.seed(100)
        # Filtered forward reads
        errF <- dada2::learnErrors(filtF, nread=1e3, multithread=as.numeric(threads))
        # Filtered reverse reads
        errR <- dada2::learnErrors(filtR, nread=1e3, multithread=as.numeric(threads))

        # Visualize the estimated error rates
         ggplot2::ggsave(error_ratesF_png, dada2::plotErrors(errF, nominalQ=TRUE), device = "png")
         ggplot2::ggsave(error_ratesR_png, dada2::plotErrors(errR, nominalQ=TRUE), device = "png")
         
         saveRDS(errF, error_ratesF_rds)
         saveRDS(errR, error_ratesR_rds)

        ''')

    errF = robjects.globalenv["errF"]
    errR = robjects.globalenv["errR"]

if __name__ == "__main__":
        main()
